### PR TITLE
Implement player movement functions

### PR DESCRIPTION
**Pull request by elmarsan**
## Changes

- Implement a bit of player movement functions
- Implement CR quickturn

### CODE REVIEW

File: src/player.cpp [hunk 1]
```diff
@@ -181,12 +193,346 @@ namespace openre::player
...
+        interop::writeJmp(0x4D9D20, pl_move);
+        interop::writeJmp(0x4DAE70, player_mv_rotate);
+        interop::writeJmp(0x004DA6C0, pl_br_03);
+        init_quickturn_move();
+    }
+
+    bool is_aiming()
+    {
+        return (
+            check_flag(FlagGroup::Status, FG_STATUS_PLAYER) && check_flag(FlagGroup::Status, FG_STATUS_GAMEPLAY)
```
Comment (IntelOrca) [inline src/player.cpp:526]: How come you are checking player flag? That just says whether you are leon or claire.

File: src/player.cpp [hunk 2]
```diff
@@ -181,12 +193,346 @@ namespace openre::player
...
+            else
+            {
+                *(uint32_t*)&player->routine_0 = 1;
+            }
+            break;
+        }
+
+        if (*player->pNow_seq & 0x4000)
+        {
+            snd_se_walk(0, 3 * ((*player->pNow_seq >> 13) & 1) + 4, player);
```
Comment (IntelOrca) [inline src/player.cpp:489]:
```
int player_get_floor_sound(PlayerEntity* player)
{
    return (*player->pNow_seq >> 13) & 1;
}
```

```suggestion
            snd_se_walk(0, 4 + (player_get_floor_sound(player) * 3), player);
```

File: src/player.cpp [hunk 3]
```diff
@@ -181,12 +193,346 @@ namespace openre::player
...
+            gPlayerEntity.spd.z = 0;
+            gPlayerEntity.move_no = *reinterpret_cast<uint32_t*>(&gPlayerEntity.d_life_u) + 458752;
+            gGameTable.fg_status &= 0xffffffC0;
+        }
+
+        joint_move(&gPlayerEntity, gPlayerEntity.pKan_t_ptr, gPlayerEntity.pSeq_t_ptr, 512);
+
+        if (*gPlayerEntity.pNow_seq & 0x4000)
+        {
+            snd_se_walk(0, 3 * ((*gPlayerEntity.pNow_seq >> 13) & 1) + 4, &gPlayerEntity);
```
Comment (IntelOrca) [inline src/player.cpp:354]: Can use same function suggestion below

File: src/player.cpp [hunk 4]
```diff
@@ -181,12 +193,346 @@ namespace openre::player
...
+            gPlayerEntity.routine_2 = 1;
+            gPlayerEntity.spd.x = 0;
+            gPlayerEntity.spd.z = 0;
+            gPlayerEntity.move_no = *reinterpret_cast<uint32_t*>(&gPlayerEntity.d_life_u) + 458752;
+            gGameTable.fg_status &= 0xffffffC0;
+        }
+
+        joint_move(&gPlayerEntity, gPlayerEntity.pKan_t_ptr, gPlayerEntity.pSeq_t_ptr, 512);
+
+        if (*gPlayerEntity.pNow_seq & 0x4000)
```
Comment (IntelOrca) [inline src/player.cpp:352]: Add constants for now_seq stuff

File: src/player.cpp [hunk 5]
```diff
@@ -181,12 +193,346 @@ namespace openre::player
...
+                gPoisonTimer = 30;
+                if (player->life > 1)
+                {
+                    player->life--;
+                }
+            }
+        }
+
+        gMoveTypeTable[player->routine_0](player);
+        pl_neck(0x1B58u, 1500);
```
Comment (IntelOrca) [inline src/player.cpp:298]:
```suggestion
        pl_neck(7000, 1500);
```

File: src/player.cpp [hunk 6]
```diff
@@ -181,12 +193,346 @@ namespace openre::player
...
+        gMoveBrTable[player->routine_1](player, gGameTable.g_key, gGameTable.key_trg);
+        gMoveMvTable[player->routine_1](player, player->pKan_t_ptr, player->pSeq_t_ptr);
+    }
+
+    // 0x004DA6C0
+    static void pl_br_03(PlayerEntity* player, uint32_t key, uint32_t key_trg)
+    {
+        static uint8_t tbl[] = { 0x28, 0x10, 0x10 };
+
+        if (!player->routine_2 || !player->move_cnt)
```
Comment (IntelOrca) [inline src/player.cpp:394]: Not a fan of using `!` unless it is a boolean or bit mask.

File: src/player.cpp [hunk 7]
```diff
@@ -181,12 +193,346 @@ namespace openre::player
...
+            {
+                player->d_life_u = 2;
+            }
+            if (t != player->d_life_u)
+            {
+                player->routine_2 = 0;
+            }
+        }
+        // back is pressed
+        if (key & 4)
```
Comment (IntelOrca) [inline src/player.cpp:416]: Constants could be used, then comments removed

File: src/player.cpp [hunk 8]
```diff
@@ -181,12 +193,346 @@ namespace openre::player
...
+            if (key & 8)
+            {
+                player->cdir.y -= tbl[player->d_life_u];
+            }
+
+            // run/cancel
+            if (key_trg & 0x0200)
+            {
+                // trigger quickturn
+                *(uint32_t*)&player->routine_0 = 0xC01;
```
Comment (IntelOrca) [inline src/player.cpp:433]: I think the individual routine fields should be set rather than doing this.

File: src/player.cpp [hunk 9]
```diff
@@ -181,12 +193,346 @@ namespace openre::player
...
+
+    // 0x004D97B0
+    static void player_move(PlayerEntity* player)
+    {
+        if (gGameTable.fg_stop < 0)
+        {
+            return;
+        }
+        gGameTable.word_989EEE &= 0xE0;
+        gGameTable.fg_status &= ~0x80;
```
Comment (IntelOrca) [inline src/player.cpp:252]: Could use `set_flag` here

File: src/player.cpp [hunk 10]
```diff
@@ -181,12 +193,346 @@ namespace openre::player
...
+        {
+            return;
+        }
+        else
+        {
+            gPlayerEntity.routine_2 = 1;
+            gPlayerEntity.spd.x = 0;
+            gPlayerEntity.spd.z = 0;
+            gPlayerEntity.move_no = *reinterpret_cast<uint32_t*>(&gPlayerEntity.d_life_u) + 458752;
+            gGameTable.fg_status &= 0xffffffC0;
```
Comment (IntelOrca) [inline src/player.cpp:347]: Could use `set_flag` here

File: src/player.cpp (hunk 1 above)
Comment (elmarsan) [reply src/player.cpp:526]: Good point, I don't remember why I did that.

After removing that flag I see some errors on `is_picking_up_item` function, as is not working yet and we do not needed for the moment I removed it. `is_aiming` seems to be correct.

**Comment by elmarsan** [comment]
I've done the changes, PTAL.

File: src/player.cpp [hunk 11]
```diff
@@ -352,7 +352,7 @@ namespace openre::player
             gPlayerEntity.spd.x = 0;
             gPlayerEntity.spd.z = 0;
             gPlayerEntity.move_no = *reinterpret_cast<uint32_t*>(&gPlayerEntity.d_life_u) + 458752;
-            gGameTable.fg_status &= 0xffffffC0;
+            set_flag(FlagGroup::Status, FG_STATUS_31, true);
```
Comment (IntelOrca) [inline src/player.cpp:355]: Isn't this unsetting flags 0 to 5?

**Pull request merged**